src/executor/sort_merge_join.rs

```diff
@@ -0,0 +1,147 @@
+#![allow(dead_code)]
+use std::cmp::Ordering;
+use std::vec::Vec;
+
+use super::*;
+use crate::array::{DataChunk, DataChunkBuilder};
+use crate::types::{DataType, Row};
+/// The executor for sort merge join
+pub struct SortMergeJoinExecutor {
+    pub left_child: BoxedExecutor,
+    pub right_child: BoxedExecutor,
+    pub left_column_index: usize,
+    pub right_column_index: usize,
+    pub left_types: Vec<DataType>,
+    pub right_types: Vec<DataType>,
+}
+impl SortMergeJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut left_rows = row_stream(self.left_child);
+        let mut right_rows = row_stream(self.right_child);
+        let mut left_row;
+        let mut right_row;
+        let first_left_row = left_rows.next().await;
+        let first_right_row = right_rows.next().await;
+        if first_left_row.is_none() || first_right_row.is_none() {
+            yield Err(ExecutorError::NotNullable)?
+        } else {
+            left_row = first_left_row.unwrap()?;
+            right_row = first_right_row.unwrap()?;
+            // build
+            let data_types = self.left_types.iter().chain(self.right_types.iter());
+            let mut builder = DataChunkBuilder::new(data_types, PROCESSING_WINDOW_SIZE);
+
+            loop {
+                match compare_row(
+                    &left_row,
+                    &right_row,
+                    self.left_column_index,
+                    self.right_column_index,
+                ) {
+                    Ordering::Equal => {
+                        let values = left_row.into_iter().chain(right_row.clone().into_iter());
+                        if let Some(chunk) = builder.push_row(values) {
+                            yield chunk;
+                        }
+                        if let Some(row) = left_rows.next().await {
+                            left_row = row?;
+                        } else {
+                            break;
+                        }
+                    }
+                    Ordering::Greater => {
+                        if let Some(row) = right_rows.next().await {
+                            right_row = row?;
+                        } else {
+                            break;
+                        }
+                    }
+                    Ordering::Less => {
+                        if let Some(row) = left_rows.next().await {
+                            left_row = row?;
+                        } else {
+                            break;
+                        }
+                    }
+                }
+            }
+            // if rows line < PROCESSING_WINDOW_SIZE ,take rest rows out of builder
+            if let Some(chunk) = { builder }.take() {
+                yield chunk;
+            }
+        }
+    }
+}
+
+pub fn compare_row(
+    left_row: &Row,
+    right_row: &Row,
+    left_column_indexes: usize,
+    right_column_indexes: usize,
```

Comment:
Are they `index` or `indexes`?

Reply:
Let's support single index condition first. We can later support multiple join keys, or even non-equal conditions.

Reply:
... and in this case, better to rename to `index`.